### ISSUE TITLE
Sheet for avrae

### ISSUE THREAD

**Issue by NoukonNeko** [oganm/printSheetApp#2]
This is more of a feature request than a current issue, but think you're able to add support for the Avrae the D&D Discord Bot?
[Avrae](https://avrae.io/)

Avrae can import character sheets from a PDF or Google Sheet.  It looks like the PDF (the pretty version) is similar to Avrae's required PDF format, but pretty pdf doesn't import to Avrae cleanly and produces an invalid character sheet error.  Here's the PDF that Avrae uses:
[PDF](http://andrew-zhu.com/avrae/PDFSheet.pdf)

The other route is to add an option to export to the the google sheet format:
[GSheets](https://docs.google.com/spreadsheets/d/1etrBJ0qCDXACovYHUM4XvjE0erndThwRLcUQzX6ts8w/edit?usp=sharing)

I personally use the GSheets by manually copying over values from the 5th Edition Character Sheet Android App over, but if you app can bridge the two great tools together by handling the export/import then I'm sure the community will find this app more valuable.

**Comment by oganm** [comment]
Is the PDF you sent me from dicecloud or DnD beyond? I get a different PDF from DnDbeyond and dicecloud seems to be integrated through the API. If there is a software that automatically fills this PDF I need to know how exactly are the fields are filled.

Specifically I need to know

- if the backstory and spell description bits are important
- what is the format for the weapons after the third one
- will it go crazy if I write `d6+-1` as weapon damage
- how is expertise coded. Doesn't seem to be way to tell the sheet to add double or half proficiency
- is the spellcasting class field actually important and why isn't eldritch knight there as an option

If there is a software that fills this PDF, you can send me a functioning character that is preferably a wizard  (to see how spell slots ) with rogue levels (to see how expertise is encoded and more than 3 weapons, then I should be able to replicate it with relative ease on a spare time (about to go for a conference so it might have to wait for a bit).

Googlesheet is also possible but that'll be more work. That sheet also doesn't seem to have expertise support. I'd also need to know if it can read the weapon table after the 5th one.

**Comment by NoukonNeko** [comment]
The PDF is not from dicecloud or DnD beyond.  It seems to be a PDF from Avrae itself.  There doesn't seem to be a software that fills the PDF, the user has to fill in the PDF manually.  The PDF works like Google Sheets that auto fills most of the sheet after the user manually fills in base stats, checks on expertise, fill in weapons, and fill in spells.

Answering your specific questions in the same order:
- They are not important is Avrae does import them so when the user types "!desc" it does print out what was written on the PDF.  Spell description are not important since Avrae already has descriptions of all spells
- There doesn't appear to be a format after the third one, seems like Avrae just ignores everything after the weapon table and the user needs to manually add it in themselves after the import.
- It accepts that format, I'll link an image of the attack results for you at the end
- Expertise doesn't seem to be supported, again it seems like the user has to manually adjust after the import
- Spell casting field just needs to know the names of spells, user can type "!spellbook" to see their spells and cast them using "!cast".  I'll link an example of that too.  I haven't played enough DnD to know the ins and outs of an Eldritch knight.

I'll attach some of the things I mention that might be helpful to you:
[charSheetPretty.pdf](https://github.com/oganm/printSheetApp/files/2164671/charSheetPretty.pdf)
[TorgaardAvrae.pdf](https://github.com/oganm/printSheetApp/files/2164684/TorgaardAvrae.pdf)
![torgaardattacks](https://user-images.githubusercontent.com/17016993/42299256-0e4d9568-7fbf-11e8-9e74-5bb494214744.png)
![torgaardcasts](https://user-images.githubusercontent.com/17016993/42299258-10b8b3e6-7fbf-11e8-9d94-be5bd991ba69.png)

What I basically did was export to a pretty character sheet then copy the values over to Avrae PDF format.  Let me know if you have anymore questions

**Comment by oganm** [comment]
I have decided to go for the googlesheet support since it would also add another useful output type even for non avrae users (it also has expertise and jack of all trades support). If it gets annoying to deal with I'll just have the pdf version up

**Comment by oganm** [comment]
A test version is up here

http://oganm.com/shiny/testing/?valid=TRUE

it is a little slow to create currently, I am working on that. but let me know if the output file has other problems

**Comment by oganm** [comment]
~~I just broke HP and don't have to time before sleep so ignore that for now~~

**Comment by NoukonNeko** [comment]
Sorry it took me some time to test it.  I asked some friends to send me their exported files so I can have more real samples, that just took a while for me to get my hands on.

The Torgaard file I sent you seem to work when converting to google sheet, but any other exported files seem to to come up with this error:
Error: no applicable method for 'filter_' applied to an object of class "NULL"

I am able to convert to a pretty PDF sheet, but not the google sheet.  I'll attach one of the exported file that would give that error:
[Spinch_Barbarian1.txt](https://github.com/oganm/prinItSheetApp/files/2212996/Spinch_Barbarian1.txt)

**Comment by NoukonNeko** [comment]
If I had to guess, it has something to do with caster vs non-casters

**Comment by oganm** [comment]
I think I found the problem (yep... casters vs non-casters, silly one at that) and fixed it but I can't test it on your file because the attachment seems to have failed. If you still see the problem, you can mail me the file at ogan.mancarci@gmail.com so I can take another look.

Otherwise what's keeping this on testing is the speed. Currently, due to absence of asynchronous operations, if someone decides to use the avrae sheet, the app will be unresponsive to everyone else for about 2 minutes. It is also true for other operations but they take much less time so that's not that big of an issue. So I'll have to do one of these things:
* speed it up significantly
* implement asynchronous operations 
* decide that I don't care about this because two different people using this within 2 minutes of each other is a rare event based on my usage data (3% of all entries by different individuals took place after less than 5 minutes from a previous one)

Till one of those happens, this will live at the testing address. Meanwhile, if you make a habit of using this, you might want to copy the output file as currently these just clutter my own drive so I might mass delete them at some point. I added a warning to the output to reflect this.

**Comment by NoukonNeko** [comment]
Yeah I been making my own copies after every conversion.  Thanks for the warning though.

A couple of other bugs I noticed:
- Max HP has been missing on recent conversions, it was working fine before, but seems to be blank now. Current HP seems correct
- Features and Traits seems to be repeating the last feature or trait multiple times until it fills the whole section?

**Comment by oganm** [comment]
both should be fixed

**Comment by nmartucci** [comment]
I seem to be having issues assigning a damage type to my snippet. Is there a specific argument call tag for assigning that?

**Comment by oganm** [comment]
The damage types exported to the google drive document doesn't work? Do you know what it should look like?

**Comment by nmartucci** [comment]
Whta I'm trying to do is create a snippet for the Druid Circle of the Spores feature "Symbiotic Entity" Where with a melee attack it does 1d6 poison dmg.

what I've made so far is [ !snippet sym -d 1d6 ] but it will only reflect the first attack's damage type

**Comment by oganm** [comment]
Are you asking for a way to make a snippet in avrae or does this have anything to do with the google drive sheet?

Because while this issue is named "Avrae support", it isn't actually aimed for general Avrae support. Its aimed to encourage me to include Avrae compatibility on my site.

I'd like to help you but I don't use Avrae myself so don't know what might be going wrong. You probably want to go to their [discord](https://discordapp.com/invite/pQbd4s6) for that

**Comment by oganm** [comment]
All this time I thought I failed to implement asynchronous programming for some reason but it turns out it just doesn't work on my server. Currently have no idea why. I managed to make things work in a local machine with mostly identical conditions.

Potential problems
* **Memory:** Server has very limited RAM. It is possible that the `future` is trying to open a new session but failing to do so, making everyone wait. However I feel like this should have been raising some flags. There's nothing in the logs that may indicate something like this might be happening. I also don't seem to be that low on memory to prompt failure to branch out.
* **rstudio and server installations:** Both of these are up to date. In fact they are more up to date than the local machine that manages to get everything working. They are also independent from each other so the problem should like in R or other dependencies.
* **R or other dependencies:** I updated all packages that async programming relies on both on my local machine and on the server. Everything still works fine on the local machine but not the server. 

If I fail to find out what is going wrong, I might go nuclear and open a new droplet to re-install everything from scratch. This requires a significant activation energy as it is quite a bit of work

**Comment by oganm** [comment]
Problem solved and a working version pushed. I will delete the test version shortly. You can now get your avrae files from the main https://oganm.github.io/printSheetApp/ link

**Comment by oganm** [comment]
Avrae has a new sheet [v2](https://docs.google.com/spreadsheets/d/1ApmbXHTln99fPTUpanyQRTXNzXbQ8UBTt3Uq8xInQKw/edit#gid=359784640). Also googlesheets api v3 is shutting down on March 2020 so will have to switch the code anyway. Otherwise the current version of the sheet works fine

Edit: Actually. Don't switch to v2. Seems like it has many autogeneration options which is not needed for our purposes. The manual version is also updated so I'll probably switch to [1.4](https://docs.google.com/spreadsheets/d/1etrBJ0qCDXACovYHUM4XvjE0erndThwRLcUQzX6ts8w/edit#gid=1750226729) instead
